pkg/pillar/containerd/oci.go

```diff
@@ -535,3 +531,24 @@ func (s *ociSpec) UpdateEnvVar(envVars map[string]string) {
 		}
 	}
 }
+
+// UpdateShimVMDevices grants full access to file devices for shim VM OCI specs
+func (s *ociSpec) UpdateShimVMDevices() {
+	if len(s.Get().Process.Args) == 0 {
+		return
+	}
+
+	ctrCmd := s.Get().Process.Args[0]
+
+	// default OCI specs have all devices being denied by default, check for a
+	// shim VM loader (qemu/xen) and enable full access to file devices on
+	// these cases. Later on we may need to get more fine-grained.
+	if strings.Contains(ctrCmd, "qemu-system") ||
+		strings.Contains(ctrCmd, "xen-start") {
```

Comment:
Why is this check needed? If it gets called from `kvm.go` or `xen.go`, doesn't it then always need those cgroups access?

Reply:
I think the reason is that later it will be also called from some other code)

Reply:
But perhaps we can have some dependency injection here instead of checking if some string contains `qemu`?

Reply:
For KVM and XEN is actually not needed, the check is done because of the Load function, when we load the OCI from a file we don't know which kind of container is, so we must check if it uses shimVM or not.... the idea with the UpdateShimVMDevices() is to follow the same approach for other points of the OCI that are customized by us. In a nutshell, we extend containerd's OCI struct to add custom fields that we need and then we have UpdateXXXX functions that changes the OCI as needed, for instance UpdateVifList(), UpdateFromDomain(), 
, UpdateFromVolume(), UpdateMounts(), UpdateEnvVar(), the idea was to create UpdateShimVMDevices() and just use it for both hypervisors Setup() and OCI Load()....

Reply:
Like what? I'm just curious what it could look like.

Reply:
Do you all guys ignore the OKR Call??? Shame on us!

Reply:
@christoph-zededa , regarding the string checking, the only thing that really defines a shimVM for us is the process loader. By checking only the arguments I can also keep the backward compatibility with all existent applications, i.e., without requiring any changing to existent OCI files....

Reply:
> Like what? I'm just curious what it could look like.

A flag.

@rene Where are those OCI files stored?

I expected to see those under `find /persist/containerd-system-root/io.containerd.runtime.v2.task/eve-user-apps/`

Reply:
> > Like what? I'm just curious what it could look like.
> 
> A flag.

Flag where? How to store/set it?

Reply:
That's a bit hard to say, if
> I think the reason is that later it will be also called from some other code)

But I am sure there is a way to do so.

Reply:
You mean just a parameters between functions call, ok. I thought of something more tricky)

Reply:
I am still struggling in fully understanding how the pieces should work together, but perhaps this method just sets a boolean flag in `s *ociSpec` and then later in `Load` or `CreateContainer` you check this flag and add these capabilities.
So, if it is a shimvm, you call `UpdateShimVMDevices`, otherwise not.

@OhmSpectator Option 3 of https://golang.cafe/blog/golang-functional-options-pattern.html would be a nice trick :-)

Reply:
@christoph-zededa this functional options pattern is really cool :+1:

Reply:
🥲🥲🥲